sklearn/preprocessing/tests/test_imputation.py

```diff
@@ -386,23 +392,99 @@ def check_indicator(X, expected_imputed_features, axis):
 def test_indicator_features():
     # one feature with all missng values
     X = np.array([
-       [-1,  -1,   2,   3],
-       [4,  -1,   6,  -1],
-       [8,  -1,  10,  11],
-       [12,  -1,  -1,  15],
-       [16,  -1,  18,  19]
+        [-1,  -1,   2,   3],
+        [4,  -1,   6,  -1],
+        [8,  -1,  10,  11],
+        [12,  -1,  -1,  15],
+        [16,  -1,  18,  19]
     ])
     check_indicator(X, np.array([0, 2, 3]), axis=0)
-    check_indicator(X, np.array([0, 1, 2, 3]), axis=1)
+    check_indicator(X, None, axis=1)
 
     # one feature with all missing values and one with no missing value
     # when axis=0 the feature gets discarded
     X = np.array([
-       [-1,  -1,   1,   3],
-       [4,  -1,   0,  -1],
-       [8,  -1,   1,  0],
-       [0,  -1,   0,  15],
-       [16,  -1,   1,  19]
+        [-1,  -1,   1,   3],
+        [4,  -1,   0,  -1],
+        [8,  -1,   1,  0],
+        [0,  -1,   0,  15],
+        [16,  -1,   1,  19]
     ])
     check_indicator(X, np.array([0, 3]), axis=0)
-    check_indicator(X, np.array([0, 1, 3]), axis=1)
+    check_indicator(X, None, axis=1)
+
+
+def test_indicator_features_shape():
+
+    X1 = np.array([
+        [-1,  -1,   2,   3],
+        [4,  -1,   6,  -1],
+        [8,  -1,  10,  11],
+        [12,  -1,  -1,  15],
+        [16,  -1,  18,  19]
+    ])
+
+    X2 = np.array([
+        [-1,  -1,   1,   3],
+        [4,  -1,   0,  -1],
+        [8,  -1,   1,  0],
+        [0,  -1,   0,  15],
+        [16,  -1,   1,  19]
+    ])
+
+    n_samples, n_features = X1.shape
+    imputer = Imputer(missing_values=-1, strategy='mean', axis=0)
+    imputer_with_in = clone(imputer).set_params(add_indicator_features=True)
+    imputer.fit(X1)
+    imputer_with_in.fit(X1)
+    Xt = imputer.transform(X2)
+    Xt_with_in = imputer_with_in.transform(X2)
+    imputed_features_mask = X2[:, imputer.imputed_features_] == -1
+    n_features_new = Xt.shape[1]
+    n_imputed_features = len(imputer.imputed_features_)
+    assert_equal(Xt_with_in.shape,
+                 (n_samples, n_features_new + n_imputed_features))
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    assert_array_equal(Xt_with_in,
+                       imputer_with_in.transform(sparse.csc_matrix(X2)).A)
+    assert_array_equal(Xt_with_in,
```

Comment:
One is for csr_matrix while other is for csc_matrix. I suppose it is redundant but thought it will be better to have separate assertions for both.